cmd/loadtest/server.go

```diff
@@ -0,0 +1,162 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/mattermost/mattermost-load-test-ng/config"
+	"github.com/mattermost/mattermost-load-test-ng/loadtest"
+	"github.com/mattermost/mattermost-load-test-ng/loadtest/control"
+	"github.com/mattermost/mattermost-load-test-ng/loadtest/control/simplecontroller"
+	"github.com/mattermost/mattermost-load-test-ng/loadtest/store/memstore"
+	"github.com/mattermost/mattermost-load-test-ng/loadtest/user/userentity"
+	"github.com/mattermost/mattermost-server/mlog"
+	"github.com/spf13/cobra"
+)
+
+type Agent struct {
+	loadTests map[string]*loadtest.LoadTester
+}
+
+func (a *Agent) createLoadTestHandler(w http.ResponseWriter, r *http.Request) {
+	var config config.LoadTestConfig
+	err := json.NewDecoder(r.Body).Decode(&config)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	newSimpleController := func(id int, status chan<- control.UserStatus) control.UserController {
+		ueConfig := userentity.Config{
+			ServerURL:    config.ConnectionConfiguration.ServerURL,
+			WebSocketURL: config.ConnectionConfiguration.WebSocketURL,
+		}
+		ue := userentity.New(memstore.New(), ueConfig)
+		return simplecontroller.New(id, ue, status)
+	}
+
+	lt := loadtest.New(&config, newSimpleController)
+	a.loadTests[lt.Id] = lt
+	// start := time.Now()
+	// err = lt.Run()
+	// if err != nil {
+	// 	return err
+	// }
+
+	// mlog.Info("loadtest started")
+	// time.Sleep(60 * time.Second)
+
+	// err = lt.Stop()
+	// mlog.Info("loadtest done", mlog.String("elapsed", time.Since(start).String()))
+
+	// return err
+
+	w.Header().Set("Content-Type", "application/json")
+	_ = json.NewEncoder(w).Encode(map[string]string{"loadTestId": lt.Id})
+
+}
+
+func (a *Agent) getLoadTestById(w http.ResponseWriter, r *http.Request) (*loadtest.LoadTester, error) {
+	vars := mux.Vars(r)
+	id := vars["id"]
+	lt, ok := a.loadTests[id]
+	if !ok {
+		err := fmt.Errorf("Load test with id %s not found", id)
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return nil, err
+	}
+	return lt, nil
+}
+func (a *Agent) runLoadTestHandler(w http.ResponseWriter, r *http.Request) {
+	lt, err := a.getLoadTestById(w, r)
+	if err != nil {
+		return
+	}
+	go lt.Run()
+	w.Header().Set("Content-Type", "application/json")
+	_ = json.NewEncoder(w).Encode(map[string]string{"message": "load test started"})
```

Comment:
Do we expect to pass more data in the response that we are using json ? As a first cut, can we just do - `fmt.Fprintf(w, "load test started")` ? Or is this more about being more consistent with the status handlers ? In which case, we can just do the same fmt.Fprintf() with the json string, instead of taking the json.NewEncoder route.

Reply:
I think it's fine to use JSON but I agree that for the time being it is more than enough to return a prepared JSON string without the need to encode anything.

Reply:
I'm going to disagree here. What's the benefit of writing this with printf instead of properly encoding a structure? Almost no code-size difference, this is not a hot path and you loose ability to properly compose output.

Reply:
I guess for me it was more about readability and avoid adding unnecessary complexity but it's not a big deal. This is not the kind of detail we want to spend our time on right now so I am also super okay with keeping it as it is :+1:

Reply:
The benefit would be performance. But at this stage, shaving off a few microseconds doesn't matter :smile: So no complaints from my side.